src/jit/codegenarm64.cpp

```diff
@@ -3875,7 +3875,14 @@ void CodeGen::genSIMDIntrinsic(GenTreeSIMD* simdNode)
         simdNode->gtSIMDBaseType != TYP_SHORT && simdNode->gtSIMDBaseType != TYP_BYTE &&
         simdNode->gtSIMDBaseType != TYP_UINT && simdNode->gtSIMDBaseType != TYP_ULONG)
     {
-        noway_assert(!"SIMD intrinsic with unsupported base type.");
+        // We don't need a base type for the Upper Save & Restore intrinsics, and we may find
+        // these implemented over lclVars created by CSE without full handle information (and
```

Comment:
Hmm, is this really because of the lack of handle or perhaps CSE doesn't know how to properly deal with these nodes?

Reply:
No, I checked where it's getting created, and it first checks for a handle, in which case it calls `lvSetStruct` (which will set `lvBaseType` appropriate), but otherwise it simply sets `lvSIMDType`, and doesn't have a reasonable way to determine the "base type". https://github.com/dotnet/coreclr/blob/master/src/jit/optcse.cpp#L2044

Reply:
It looks to me that the problem is not the handle, SIMD nodes won't ever get a handle anyway. If SIMD expressions are to be CSEd then the CSE code needs to recognize SIMD nodes and copy `gtSIMDBaseType` to the `LclVarDsc`.

I suppose the handle problem appears for SIMD `GT_IND` nodes as those have neither a handle nor a "base type". Mmm... that's interesting. Does this mean that SIMD typed `GT_IND` nodes are actually undesirable and that they should in fact be `GT_OBJ` nodes so we have the handle?

And now it looks like I may need to add the SIMD base type to my `ClassLayout` thing...

Reply:
This needs further thought (which I hope to be able to do once 3.0 is stable), but:
- I *do* think it's useful to have cached struct info for SIMD types that would include the base type. For the most part this info isn't all that necessary (since it will already be specified on any intrinsics that are imported), but I think that we may want to distinguish between base types for certain move operations as it may matter for performance, in which case it would be useful post-importer.
- It might be reasonable to revert to having these be `OBJ` nodes, as long as we ensure that 1) we've addressed all the place that formerly pessimized such nodes, and 2) we reduce their size.

Reply:
>  1) we've addressed all the place that formerly pessimized such nodes, and 2) we reduce their size.

Well, at least 2) is addressed by my `ClassLayout` thing. 1) might be fun though.

But I'm looking at `gtGetStructHandleIfPresent` and it's still not very clear to me how this problem appeared. `gtGetStructHandleIfPresent` is actually capable of pulling out a struct handle out of thin air, even for SIMD `GT_IND` nodes, where it defaults to `TYP_FLOAT` as base type.

Reply:
I suspect that the real problem might be here: https://github.com/dotnet/coreclr/blob/af83b8ad6fe02b45321741e26d996ee4b738d67f/src/jit/compiler.h#L7780-L7785

Reply:
@mikedn - I suspect that you're right, and it's probably some other place where the appropriate distinction between `featureSIMD` and `supportSIMDTypes()` isn't being made. I'll look into this further, but it may still be reasonable to relax the assert in codegenarm64.cpp, since the absence of a baseType doesn't impact codegen.

Reply:
I finally spent a little time to track this down. The (obvious, once I figured it out) issue is specifically that we're "defaulting" to `TYP_FLOAT` - that is, we assume `TYP_FLOAT` without even trying to figure it out from context. So, if we have a different base type, and we haven't seen the handle for the `TYP_FLOAT` case, we return `NO_CLASS_HANDLE.
I think there's a cleaner approach to handling this that doesn't involve keeping a cache of class handles, probably in conjunction with (or as an adjunct to) the class layout work that you're doing @mikedn